azurelinuxagent/common/cgroupapi.py

```diff
@@ -241,17 +245,105 @@ class SystemdCgroupsApi(CGroupsApi):
     """
     Cgroups interface via systemd
     """
+
+    @staticmethod
+    def create_and_start_unit(unit_filename, unit_contents):
+        try:
+            fileutil.write_file("/etc/systemd/system/{0}".format(unit_filename), unit_contents)
+            shellutil.run_get_output("systemctl daemon-reload")
+            shellutil.run_get_output("systemctl start {0}".format(unit_filename))
+        except Exception as e:
+            logger.warn("Failed to create and start {0}. Error: {1}".format(unit_filename, ustr(e)))
+
+    @staticmethod
+    def _get_extensions_slice_root_name():
+        return "system-{0}.slice".format(EXTENSIONS_ROOT_CGROUP_NAME)
+
+    @staticmethod
+    def _get_extension_slice_name(extension_name):
+        return "system-walinuxagent.extensions-{0}.slice".format(extension_name.replace('-', '_'))
+
+    @staticmethod
+    def _get_extension_scope_name(extension_name):
+        return extension_name.replace('-', '_')
+
     def create_agent_cgroups(self):
-        pass
+        try:
+            cgroup_unit = None
+            cgroup_paths = fileutil.read_file("/proc/self/cgroup")
+            for entry in cgroup_paths.splitlines():
+                fields = entry.split(':')
+                if fields[1] == "name=systemd":
+                    cgroup_unit = fields[2].lstrip(os.path.sep)
+
+            cpu_cgroup_path = os.path.join(CGROUPS_FILE_SYSTEM_ROOT, 'cpu', cgroup_unit)
+            memory_cgroup_path = os.path.join(CGROUPS_FILE_SYSTEM_ROOT, 'memory', cgroup_unit)
+
+            return [cpu_cgroup_path, memory_cgroup_path]
+        except Exception as e:
+            logger.warn("Failed to get paths of agent's cgroups. Error: {0}".format(ustr(e)))
 
     def create_extension_cgroups_root(self):
-        pass
+        unit_contents = """[Unit]
+Description=Slice for WALinuxAgent extensions
+DefaultDependencies=no
+Before=slices.target
+Requires=system.slice
+After=system.slice"""
+        unit_filename = self._get_extensions_slice_root_name()
+        self.create_and_start_unit(unit_filename, unit_contents)
+        logger.info("Created and started {0}".format(unit_filename))
 
     def create_extension_cgroups(self, extension_name):
-        pass
+        # TODO: revisit if we need this code now (not used until we interact with the D-bus API and run the
+        #  extension scopes within their designated slice)
+        unit_contents = """[Unit]
+Description=Slice for extension {0}
+DefaultDependencies=no
+Before=slices.target
+Requires=system-{1}.slice
+After=system-{1}.slice""".format(extension_name, EXTENSIONS_ROOT_CGROUP_NAME)
+        unit_filename = self._get_extension_slice_name(extension_name)
+        self.create_and_start_unit(unit_filename, unit_contents)
+        logger.info("Created and started {0}".format(unit_filename))
```

Comment:
since create_and_start_unit swallows all errors, any failure will go undetected (e.g. the info message will seem to indicate success). you probably want to raise an exception in create_and_start_unit and let the configurator class deal with it

Reply:
Good catch, fixed.

Reply:
? create_and_start_unit still has a catch all